File: 1095. Find in Mountain Array/findInMountainArray.py
# """
# This is MountainArray's API interface.
# You should not implement it, or speculate about its implementation
# """
#class MountainArray:
#    def get(self, index: int) -> int:
#    def length(self) -> int:

class findInMountainArray:
    def findMountainTop(self, mountain_arr: 'MountainArray'):
        left = 0
        right = mountain_arr.length() - 1
        while left < right:
            mid = left + (right - left) // 2
            midNum = mountain_arr.get(mid)
            if midNum < mountain_arr.get(mid + 1):
                left = mid + 1
            else:
                right = mid
        return left
    
    def findAscendPath(self, target, mountain_arr, top):
        left = 0
        right = top
        while left < right:
            mid = left + (right - left) // 2
            if mountain_arr.get(mid) < target:
                left = mid + 1
            else:
                right = mid

        return left if mountain_arr.get(left) == target else -1

    def findDescendPath(self, target, mountain_arr, top):
        left = top + 1
        right = mountain_arr.length() - 1
        while left < right:
            mid = left + (right - left) // 2
            if mountain_arr.get(mid) > target:
                left = mid + 1
            else:
                right = mid
        return left if mountain_arr.get(left) == target else -1

    def solution(self, target: int, mountain_arr: 'MountainArray') -> int:
        top = self.findMountainTop(mountain_arr)
        res = self.findAscendPath(target, mountain_arr, top)
        if res != -1:
            return res
        res = self.findDescendPath(target, mountain_arr, top)
        return res

File: 1095. Find in Mountain Array/findInMountainArray.java
/**
 * // This is MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface MountainArray {
 *     public int get(int index) {}
 *     public int length() {}
 * }
 */
 
class findInMountainArray {
    public int findMountainTop(MountainArray mountainArr, int size) {
        int left = 0, right = size - 1, mid = 0;
        while (left < right) {
            mid = left + (right - left) / 2;
            if (mountainArr.get(mid) < mountainArr.get(mid + 1)) {
                left = mid + 1;
            }
            else {
                right = mid;
            }
        }
        return left;
    }

    public int findInAscendPath(int target, MountainArray mountainArr, int left, int right) {
        int mid = 0;
        while (left < right) {
            mid = left + (right - left) / 2;
            if (mountainArr.get(mid) < target) {
                left = mid + 1;
            }
            else {
                right = mid;
            }
        }
        if (mountainArr.get(left) == target) {
            return left;
        }
        return -1;
    }

    public int findInDescendPath(int target, MountainArray mountainArr, int left, int right) {
        int mid = 0;
        while (left < right) {
            mid = left + (right - left) / 2;
            if (mountainArr.get(mid) > target) {
                left = mid + 1;
            }
            else {
                right = mid;
            }
        }
        if (mountainArr.get(left) == target) {
            return left;
        }
        return -1;
    }

    public int solution(int target, MountainArray mountainArr) {
        int size = mountainArr.length();
        int top = findMountainTop(mountainArr, size);
        int res = findInAscendPath(target, mountainArr, 0, top);
        if (res != -1) {
            return res;
        }
        res = findInDescendPath(target, mountainArr, top + 1, size - 1);
        return res;
    }
}